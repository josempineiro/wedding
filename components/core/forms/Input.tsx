import cn from "classnames";
import { forwardRef, ForwardedRef } from "react";

declare module "react" {
  function forwardRef<T, P = {}>(
    render: (props: P, ref: React.Ref<T>) => React.ReactElement | null
  ): (props: P & React.RefAttributes<T>) => React.ReactElement | null;
}

interface Props<T>
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, "onChange"> {
  onChange?: (value: T, event: React.ChangeEvent<HTMLInputElement>) => void;
}

function Input<T = string>(
  { className, onChange, ...props }: Props<T>,
  ref: ForwardedRef<HTMLInputElement>
) {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (onChange) {
      onChange(event.target.value as T, event);
    }
  };
  return (
    <input
      className={cn([
        className,
        "text-md ring-none min-w-0 w-20 bg-transparent outline-none",
      ])}
      onChange={handleChange}
      {...props}
      ref={ref}
    />
  );
}

const ForwardedInput = forwardRef(Input);

export { ForwardedInput as Input };
