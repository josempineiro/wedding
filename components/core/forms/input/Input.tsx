import cn from "classnames";
import { forwardRef, ForwardedRef } from "react";

declare module "react" {
  function forwardRef<T, P = {}>(
    render: (props: P, ref: React.Ref<T>) => React.ReactElement | null
  ): (props: P & React.RefAttributes<T>) => React.ReactElement | null;
}

export interface InputProps<T>
  extends Omit<
    React.InputHTMLAttributes<HTMLInputElement>,
    "onChange" | "children"
  > {
  onChange: (value: T, event: React.ChangeEvent<HTMLInputElement>) => void;
}

function Input<T = string>(
  { className, onChange, ...props }: InputProps<T>,
  ref: ForwardedRef<HTMLInputElement>
) {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (onChange) {
      onChange(event.target.value as T, event);
    }
  };
  return (
    <input
      ref={ref}
      role="textbox"
      className={cn([
        className,
        "text-md font-medium ring-none min-w-0 w-auto bg-transparent outline-none",
      ])}
      onChange={handleChange}
      {...props}
    />
  );
}

const ForwardedInput = forwardRef(Input);

export { ForwardedInput as Input };
