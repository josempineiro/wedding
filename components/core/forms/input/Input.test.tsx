import { render, screen, fireEvent } from "@testing-library/react";
import { Input } from "@/components/core/forms/input/Input";

describe("Input", () => {
  it("renders correctly", () => {
    const { getByRole } = render(<Input />);
    const input = getByRole("textbox");
    expect(input).toBeInTheDocument();
    expect(input).toMatchSnapshot();
  });

  it("calls onChange handler when text is entered", () => {
    const handleChange = jest.fn();
    render(<Input onChange={handleChange} />);
    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "test" } });
    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toHaveBeenCalledWith("test", expect.any(Object));
  });
});
