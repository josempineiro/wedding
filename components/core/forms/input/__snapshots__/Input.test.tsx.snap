// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<input
  class="text-md ring-none min-w-0 w-auto bg-transparent outline-none"
  role="textbox"
/>
`;
